server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java

```diff
@@ -535,50 +613,248 @@ protected Releasable drainSync() {
 
     @Override
     public void trimUnreferencedReaders() throws IOException {
-        // clean up local translog files and updates readers
-        super.trimUnreferencedReaders();
+        trimUnreferencedReaders(false, true);
+    }
+
+    // Visible for testing
+    protected void trimUnreferencedReaders(boolean indexDeleted, boolean trimLocal) throws IOException {
+        if (trimLocal) {
+            // clean up local translog files and updates readers
+            super.trimUnreferencedReaders();
+        }
+
+        // Update file tracker to reflect local translog state
+        Optional<Long> minLiveGeneration = readers.stream().map(BaseTranslogReader::getGeneration).min(Long::compareTo);
+        if (minLiveGeneration.isPresent()) {
+            List<String> staleFilesInTracker = new ArrayList<>();
+            for (String file : fileTransferTracker.allUploaded()) {
+                if (file.endsWith(TRANSLOG_FILE_SUFFIX)) {
+                    long generation = Translog.parseIdFromFileName(file);
+                    if (generation < minLiveGeneration.get()) {
+                        staleFilesInTracker.add(file);
+                        staleFilesInTracker.add(Translog.getCommitCheckpointFileName(generation));
+                    }
+                }
+                fileTransferTracker.delete(staleFilesInTracker);
+            }
+        }
 
         // This is to ensure that after the permits are acquired during primary relocation, there are no further modification on remote
         // store.
         if (startedPrimarySupplier.getAsBoolean() == false || pauseSync.get()) {
             return;
         }
 
+        // This is to fail fast and avoid listing md files un-necessarily.
+        if (indexDeleted == false && RemoteStoreUtils.isPinnedTimestampStateStale()) {
```

Comment:
This block is just a short-circuit mechanism. Even if we skip this, filtering based on age and based on pinned timestamps is going to happen. And as a part of  `RemoteStoreUtils.filterOutMetadataFilesBasedOnAge`, we also consider last successful fetch timestamp. So, skipping the check is safe. Actually, we can always skip it but this was added to avoid fetch of metadata files where we already know the fetch timestamp is stale and the next GC will take care of deletion.

Why is it required to skip this check?
- Because the next GC may not happen for deleted index.